### ISSUE TITLE
Stdlib should contain a lazy visualization for text

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4907]
This task is automatically imported from the old Task Issue Board and it was originally created by Michael Mauderer.
Original issue is [here](https://www.pivotaltracker.com/story/show/183217714).
<hr />

## Why

**As** Enso user
**I want** to be able to preview the output of expressions that are very very very large without slowing down the IDE
**So that** I can work with and explore large data objects

## Acceptance Criteria

```gherkin
Scenario: 
Given a example node that produces a huge text output
Then I can see the correct text visualization without IDE slow-down.
```

## Notes
This might require changes to the scrolling behaviour similar to what will be done in #182608041, and we might want to schedule these tasks sequentially to avoid duplicating work.

## Blockers:
[#183453449](https://www.pivotaltracker.com/story/show/#183453449) resolved
[#183453466](https://www.pivotaltracker.com/story/show/#183453466) resolved

## Comments:
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-15):

**Progress:** Implemented Wojciech second round review feedback for the PR "
Show Visualization Preview when Selecting Item on Searcher". Started work on "Lazy Text" and discussing with Dmitry issues with the current implementation of Lazy Tables (as any functionality or issues  there might well be relevant to my task) It should be finished by 2022-09-23.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Continue with helping Dmitry and clarifying some technical discussions about how lazy text should handle the unknown width of text (which can only be known once the font is known). (Enso Bot - Sep 15, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-18):

**Progress:** Got QA approval for and merged PR "
Show Visualization Preview when Selecting Item on Searcher". Updated the PR for import handling, but spotted some issues with the engine that need to be discussed and resolved before this can be merged. Discussing with Dmitry work on Lazy Tables and found a few open questions for "Lazy Text" that need to be answered (see Discord Thread on new IDE channel). It should be finished by 2022-09-23.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Resolve open discussions, investigate engine issue  for import review. Follow up on a few reported regressions. (Enso Bot - Sep 18, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for yesterday (2022-09-19):

**Progress:** Found that the engine issues that I spotted in when testing the PR about import handling are unrelated to the PR, so re-released the PR for review. Followed up on some regressions and did testing. Discussed some implementation details about "Lazy Text" and started looking at the implementation of the text splitting on the Enso side. It should be finished by 2022-09-23.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Some planning and task estimation. Figure out why node reversion is not fixed on develop, when it was fixed on my branch (and even video recorded). Make some progress on "Lazy Text" (Enso Bot - Sep 20, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-20):

**Progress:** Investigated why the edit-reversion-edit-fix is not working on develop. Not sure yet what has changed, and trying to hunt down the working version. Also did some planning with Adam and looked into what work I did for the updated doc panel about a year ago when we started working on this task the first time. It should be finished by 2022-09-23.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Re-fix the edit reversion bug. Continue work on "Lazy Text" (Enso Bot - Sep 20, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-21):

**Progress:** Found that the edit-reversion-fix is working for some nodes, but not all. I'm not quite sure yet, what the difference between the nodes is, but this seems a big clue as to what is going on. Implemented Adams review feedback on the import handling PR.  Not much progress on the Lazy Text unfortunately, but I should be back on that tomorrow. It should be finished by 2022-09-23.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Find solution for extended fix the for the edit reversion bug. Continue work on "Lazy Text". (Enso Bot - Sep 21, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-22):

**Progress:** Created PR for edit-reversion-fix. Implemented Adam's second round review feedback, got QA approval, waiting on CI now. Continue work on "Lazy Text" extending some functionality related to the scrolling to gather line numbers to calculate string positions. It should be finished by 2022-09-23.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Continue work on "Lazy Text" open PR if things go well. (Enso Bot - Sep 22, 2022)
<hr />
**Michael Mauderer** reports a new **🔴 DELAY** for today (2022-09-25):

**Summary:** There is **4 days** delay in implementation of the [Stdlib should contain a lazy visualization for text (#183217714) task](https://www.pivotaltracker.com/story/show/183217714).
It **will cause 0 days delay** for the delivery of this weekly plan.

2 Days weekend

**Delay Cause:** Progress on implementation was slow, due to fixing another bug/regression from a previous task. (Enso Bot - Sep 25, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-25):

**Progress:** Continued implementation of the lazy text vis. Started an implementation of Enso preprocessor and when hooking it up to the scrolling I ran into an issue with the JS callbacks which seem to disable scrolling when trying to get scroll status. It should be finished by 2022-09-27.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Fix scrolling callback issue. Finish up remaining parts of the implementation (manual end-to-end testing of horizontal scrolling and implement vertical scrolling). (Enso Bot - Sep 25, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for yesterday (2022-09-26):

**Progress:** Continued working implementation of lazy text, implementing some Enso preprocessor code. Ran into some issues building the engine. It should be finished by 2022-09-27.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Talk to Dmitry about Engine workflow and state of Lazy Table. Continue on lazy text. (Enso Bot - Sep 27, 2022)
<hr />
**Michael Mauderer** reports a new **🔴 DELAY** for today (2022-09-27):

**Summary:** There is **3 days** delay in implementation of the [Stdlib should contain a lazy visualization for text (#183217714) task](https://www.pivotaltracker.com/story/show/183217714).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** After some discussion with Dmitry and Wojciech we decided to pivot the implementation to use Grid View, scrapping some of the existing work and unifying the implementation of Text Vis and Table Vis. (Enso Bot - Sep 27, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for yesterday (2022-09-26):

**Progress:** Resolved issues with the engine build, thanks to Dmitry. Discussed state of work of Lazy Table Vis with Dmitry and had a follow-up chat with Wojciech where we decided to pivot the implementation to use Grid View. It should be finished by 2022-09-30.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Test out Grid View with text Dom elements in a Demo scene to see whether there are any unforeseen issues with this approach. If that works, start wrapping this as a Text Grid Visualization, which will serve as the basis for the text visualization (and later the Table Visualisation). (Enso Bot - Sep 27, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-27):

**Progress:** Resolved issues with the engine build, thanks to Dmitry. Discussed state of work of Lazy Table Vis with Dmitry and had a follow-up chat with Wojciech where we decided to pivot the implementation to use Grid View. It should be finished by 2022-09-30.

**Next Day:** Next day I will be working on the [##182634001 task](https://www.pivotaltracker.com/story/show/182634001). Test out Grid View with text Dom elements in a Demo scene to see whether there are any unforeseen issues with this approach. If that works, start wrapping this as a Text Grid Visualization, which will serve as the basis for the text visualization (and later the Table Visualisation). (Enso Bot - Sep 27, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-28):

**Progress:** Started working on demo scene for Text Grid Visualization. So far, no problems. Need to test a few more things tomorrow, but if nothing bad comes up, there should be a PR for Lazy Text Vis on Friday/Monday It should be finished by 2022-09-30.

**Next Day:** Next day I will be working on the [##183217714 task](https://www.pivotaltracker.com/story/show/183217714). Continue with work on Text Grid Visualization demo scene. (Enso Bot - Sep 28, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-09-29):

**Progress:** Continued work on demo scene, getting a bit deeper into the finer points of the GridView. This took a bit longer than hoped. It should be finished by 2022-09-30.

**Next Day:** Next day I will be working on the [##183217714 task](https://www.pivotaltracker.com/story/show/183217714). Continue with work on Text Grid Visualization demo scene. (Enso Bot - Sep 29, 2022)
<hr />
**Michael Mauderer** reports a new **🔴 DELAY** for today (2022-10-03):

**Summary:** There is **5 days** delay in implementation of the [Stdlib should contain a lazy visualization for text (#183217714) task](https://www.pivotaltracker.com/story/show/183217714).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Implementing the new visualization with the GridView took longer than expected. Part of this was getting familiar with the API and working out kinks of the implementation. Will propose a task split to better track progress and reflect current knowledge about what needs to be done. (Enso Bot - Oct 3, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for the provided date (2022-10-01):

**Progress:** Was working on creating the demo scene with a scrollable GridView that contains HTML elements. Made some progress, but working with the GridView took a bit longer than I had hoped. It should be finished by 2022-10-05.

**Next Day:** Next day I will be working on the [##183217714 task](https://www.pivotaltracker.com/story/show/183217714). Continue with work on Text Grid Visualization demo scene. Propose task split for demo scene with visualization (basic working clipping/scrolling/correct alignemtns and otehr layout and engine integration (caching of lines, updating visualization preprocessors) (Enso Bot - Oct 3, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for yesterday (2022-10-03):

**Progress:** Was working on creating the demo scene with a scrollable GridView that contains HTML elements. Was working on creating a correct layout and investigating scrolling and clipping. After planning, had a discussion with Adam about using a different GridView that does not use a clipping area for some performance gains. It should be finished by 2022-10-05.

**Next Day:** Next day I will be working on the [##183217714 task](https://www.pivotaltracker.com/story/show/183217714). Move of the scrolling grid view for a simpler one and implement my own scrolling. Add HTML clipping to the texts. (Enso Bot - Oct 4, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-10-04):

**Progress:** Continued work on the Text Grid View Demo Scene. Added more HTML based layout features. Made some good progress on scrolling and clipping. Spent some time on unblocking the rustc compiler upgrade after the big text rendering PR landed. It should be finished by 2022-10-05.

**Next Day:** Next day I will be working on the [##183217714 task](https://www.pivotaltracker.com/story/show/183217714). Some more work on the rustc upgrade, after Michal encountered some issues after upgrading the CI code base. Hopefully, finish the Text Grid View Demo scene and open PR. (Enso Bot - Oct 4, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2022-10-05):

**Progress:** Continued work on the Text Grid View Demo Scene. Started moving the demo code into our visualization wrapper, but got sidetracked by the rustc bump, so there will be a slight delay with the PR. It should be finished by 2022-10-05.

**Next Day:** Next day I will be working on the [##183217714 task](https://www.pivotaltracker.com/story/show/183217714). (will be Monday, see time off) Addressing comments from rustc bump PR and finishing up the cleanup of the Text Grid Visualization code. (Enso Bot - Oct 5, 2022)
<hr />
**Michael Mauderer** reports a new **🔴 DELAY** for yesterday (2022-10-13):

**Summary:** There is **5 days** delay in implementation of the [Stdlib should contain a lazy visualization for text (#183217714) task](https://www.pivotaltracker.com/story/show/183217714).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** 2 days weekend + 1 day working on another task + 2 days for testing and layouting taking longer than hoped. (Enso Bot - Oct 14, 2022)
<hr />
